statsmodels/tsa/vecm/vecm.py

```diff
@@ -0,0 +1,557 @@
+from math import log, pi
+import numpy as np
+from numpy import hstack, vstack
+from numpy.linalg import inv, svd
+import scipy
+import scipy.stats
+from statsmodels.tools.decorators import cache_readonly
+from statsmodels.tools.tools import chain_dot
+from statsmodels.tsa.tsatools import duplication_matrix, vec, vech, unvec
+
+import statsmodels.tsa.base.tsa_model as tsbase
+
+def mat_sqrt(_2darray):
+    u_, s_, v_ = svd(_2darray, full_matrices=False)
+    s_ = np.sqrt(s_)
+    return chain_dot(u_, np.diag(s_), v_)
+
+def _endog_matrices(endog_tot, diff_lags, deterministic):
+    # p. 286:
+    p = diff_lags+1
+    y = endog_tot
+    K = y.shape[0]
+    y_1_T = y[:, p:]
+    T = y_1_T.shape[1]
+    delta_y = np.diff(y)
+    delta_y_1_T = delta_y[:, p-1:]
+    y_min1 = y[:, p-1:-1]
+    if "lt" in deterministic:
+        y_min1 = vstack((y_min1,
+                         np.arange(T)))  # p. 299
```

Comment:
I think it's better to use two deterministic options in the code (but maybe not in the general API), one for the cointegrating relationship here, i.e. y_min1 with no extra deterministic with constant or with trend. That's D_co in Lutkepohl
http://www.jmulti.de/download/help/vecm.pdf  From section 3.3, I assume the deterministic options refer to D_co, I don't see what specified what's in the difference equation, i.e. `D`